yt/java/ytsaurus-client/src/main/java/tech/ytsaurus/client/MultiYTsaurusClient.java

```diff
@@ -306,276 +355,127 @@ public Builder setPenaltyProvider(PenaltyProvider penaltyProvider) {
-            String shortClusterName = client.getClusterName().split("\\.")[0];
-            client.externalPenalty = penaltyProvider.getPenalty(shortClusterName);
-            Duration currentPenalty = client.getPenalty();
-            if (currentPenalty.compareTo(minPenalty) < 0) {
-                minPenalty = currentPenalty;
-            }
-        }
-        return minPenalty;
-    }
-
-    private void onFinishRequest(
-            int clientId,
-            Duration effectivePenalty,
-            Duration adaptivePenalty,
-            @Nullable Throwable error
+    private final List<ClientEntry> clients;
+    private final Duration banPenalty;
+    private final Duration banDuration;
+    private final PenaltyProvider penaltyProvider;
+    private final MultiExecutorMonitoring executorMonitoring;
+
+    public MultiExecutor(
+            List<MultiYTsaurusClient.YTsaurusClientOptions> clientOptions,
+            Duration banPenalty,
+            Duration banDuration,
+            PenaltyProvider penaltyProvider,
+            MultiExecutorMonitoring executorMonitoring
     ) {
-        YTsaurusClientEntry client = clients.get(clientId);
-        boolean isCancel = (effectivePenalty.compareTo(Duration.ZERO) > 0 && (error instanceof CancellationException));
-
-        if (error == null) {
-            if (adaptivePenalty.compareTo(Duration.ZERO) > 0) {
-                synchronized (this) {
-                    client.banUntil = null;
-                    client.adaptivePenalty = Duration.ZERO;
-                }
-            }
-        } else if (!isCancel) {
-            synchronized (this) {
-                client.banUntil = Instant.now().plus(banDuration);
-                client.adaptivePenalty = client.adaptivePenalty.plus(banPenalty);
-            }
-        }
+        this.clients = clientOptions.stream().map(ClientEntry::new).toList();
+        this.banPenalty = banPenalty;
+        this.banDuration = banDuration;
+        this.penaltyProvider = penaltyProvider;
+        this.executorMonitoring = executorMonitoring;
     }
 
     <R> CompletableFuture<R> execute(Function<BaseYTsaurusClient, CompletableFuture<R>> callback) {
-        Instant now = Instant.now();
-        List<YTsaurusClientEntry> currentClients = updateAdaptivePenalty(now);
-        Duration minPenalty = getMinPenalty(currentClients);
-        List<DelayedTask<R>> delayedTasks = new ArrayList<>();
-        CompletableFuture<R> result = new CompletableFuture<>();
-
-        AtomicInteger finishedClientsCount = new AtomicInteger(0);
-
-        for (int clientId = 0; clientId < currentClients.size(); ++clientId) {
-            YTsaurusClientEntry client = currentClients.get(clientId);
-            Duration effectivePenalty = client.getPenalty().minus(minPenalty);
+        return new MultiExecutorRequestTask<>(
+                getEffectiveClientOptions(Instant.now()),
+                callback,
+                executorMonitoring
+        ).getFuture();
```

Comment:
Looks a bit weird. Maybe rename getFuture to execute (and put execution logic in that method)?

Reply:
Haven't come up with a better solution which wouldn't require big changes. MultiRequestTask is actually a smart future wrapper, so it is hard to separate creation of this future and of the task